xla/tools/multihost_hlo_runner/python_hlo_runner.cc

```diff
@@ -0,0 +1,448 @@
+  TF_ASSIGN_OR_RETURN(FunctionalHloRunner::RunningOptions running_options,
+                      RunningOptionsFromFlags(opts));
+
+  // tsl::Flags::Parse() leaves unknown flags in argv, we assume that those are
+  // HLO files to run. Note that argv[0] is the binary name and is excluded.
+  if (hlo_files.size() == 0) {
+    return absl::InvalidArgumentError("No HLO files provided.");
+  }
+
+  if (!opts.dump_output_literal_to.empty() && hlo_files.size() > 1) {
+    return absl::InvalidArgumentError(
+        "Can only dump output literal when single input file is specified.");
+  }
+
+  if (opts.gpu_client_mem_fraction < 0.0 ||
+      opts.gpu_client_mem_fraction > 1.0) {
+    return absl::InvalidArgumentError(
+        "Invalid GPU client memory fraction. Must be in range [0.0, 1.0]");
+  }
+
+  PjRtEnvironment env;
+  std::unique_ptr<HLORunnerProfiler> hlo_runner_profiler;
+  if (opts.device_type == DeviceType::kGpu) {
+    GpuClientOptions gpu_options;
+    gpu_options.node_id = opts.task_id;
+    gpu_options.num_nodes = opts.num_nodes;
+    gpu_options.enable_mock_nccl = opts.enable_mock_nccl;
+    gpu_options.allocator_config.memory_fraction = opts.gpu_client_mem_fraction;
+    TF_ASSIGN_OR_RETURN(
+        env, GetPjRtEnvironmentForGpu(
+                 opts.address, gpu_options,
+                 absl::Seconds(opts.gpu_client_initialization_timeout_sec)));
+  } else {
+    QCHECK(opts.device_type == DeviceType::kHost) << "Invalid device type";
+    TF_ASSIGN_OR_RETURN(env, GetPjRtEnvironmentForHostCpu());
+  }
+
+  CHECK(env.client != nullptr);
+  if (!opts.xla_gpu_dump_xspace_to.empty()) {
+    TF_ASSIGN_OR_RETURN(hlo_runner_profiler,
+                        HLORunnerProfiler::Create(opts.xla_gpu_dump_xspace_to,
+                                                  /*keep_xspace=*/false));
+    running_options.profiler = hlo_runner_profiler.get();
+  }
+
+  for (const auto& hlo_file : hlo_files) {
+    std::vector<ExecutionProfile> execution_profiles;
+    if (opts.profile_execution) {
+      running_options.execution_profiles = &execution_profiles;
+    }
+    if (opts.should_run) {
+      TF_RETURN_IF_ERROR(FunctionalHloRunner::LoadAndRunAndDump(
+          *env.client, GetDebugOptionsFromFlags(), preproc_options,
+          raw_compile_options, running_options, hlo_file, opts.input_format,
+          opts.dump_output_literal_to, opts.task_id));
+    } else {
+      TF_RETURN_IF_ERROR(FunctionalHloRunner::LoadAndCompile(
+          *env.client, GetDebugOptionsFromFlags(), preproc_options,
+          raw_compile_options, hlo_file, opts.input_format, opts.task_id));
+    }
+    for (int i = 0; i < execution_profiles.size(); ++i) {
+      LOG(INFO) << "## Execution time, file=" << hlo_file << " repeat=" << i
+                << " duration=" << execution_profiles[i].compute_time_ns()
+                << "ns" << std::endl;
+    }
+  }
+}
+
+absl::Status RegisterCustomCallTarget(const std::string& fn_name, nb::object fn,
```

Comment:
Isnt custom call registration already exposed at the Python level? I found `PyRegisterCustomCallTarget` in jaxlib/xla_compiler.cc. Can we reuse that here instead?

Reply:
Thought about this again. Yes, we don't want a dependency on JAX here, so I guess the little bit of duplicate code is fine. Thanks for investigating